File: queue/priority_queue/0347_top_k_frequent_element.js
/*
Given an integer array nums and an integer k, return the k most frequent 
elements. You may return the answer in any order.

k is in the range [1, the number of unique elements in the array].

It is guaranteed that the answer is unique.
*/
// https://leetcode.com/problems/top-k-frequent-elements/

// hash table + sorting
// time complexity O(n * k)
// space complexity O(n)
var topKFrequent = function (nums, k) {
  const map = new Map();
  const res = [];
  // make a dictionary of nums
  nums.forEach((num) => map.set(num, (map.get(num) || 0) + 1));
  // sorting
  const store = [...map.entries()].sort((a, b) => b[1] - a[1]);
  for (let i = 0; i < k; i++) {
    res.push(store[i][0]);
  }
  return res;
};

// counting sort
// time complexity O(n * k)
// space complexity O(n)
var topKFrequent = function (nums, k) {
  const numsMin = Math.min(...nums);
  const numsMax = Math.max(...nums);
  const count = new Array(numsMax - numsMin + 1).fill(0);
  for (let num of nums) {
    count[num - numsMin]++;
  }
  const res = [];
  while (k > 0) {
    const countMax = Math.max(...count);
    const index = count.indexOf(countMax);
    res.push(index + numsMin);
    count[index] = -Infinity;
    k--;
  }
  return res;
};

File: queue/priority_queue/0347_top_k_frequent_element.py
# Given an integer array nums and an integer k, return the k most frequent 
# elements. You may return the answer in any order.

# k is in the range [1, the number of unique elements in the array].

# It is guaranteed that the answer is unique.

# https://leetcode.com/problems/top-k-frequent-elements/

# priority queue / heap sort
import heapq
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        freq_map = dict()
        for num in nums:
            freq_map[num] = freq_map.get(num, 0) + 1
        top_k_ele = []
        for num, freq in freq_map.items():
            heapq.heappush(top_k_ele, (freq, num))
            if len(top_k_ele) > k:
                heapq.heappop(top_k_ele)
        top_k = []
        while top_k_ele:
            top_k.append(heapq.heappop(top_k_ele)[1])
        return top_k
    
# bucket sort
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        freq_map = dict()
        for num in nums:
            freq_map[num] = freq_map.get(num, 0) + 1
        buckets = [[] for _ in range(max(freq_map.values()) + 1)]
        for num in freq_map:
            buckets[freq_map[num]].append(num)
        res = []
        for i in range(len(buckets) - 1, 0, -1):
            for num in buckets[i]:
                res.append(num)
        return res[:k]

